arrow-core-data/src/main/kotlin/arrow/core/Either.kt

```diff
@@ -926,6 +926,60 @@ sealed class Either<out A, out B> : EitherOf<A, B> {
       } catch (t: Throwable) {
         fe(t.nonFatalOrThrow()).left()
       }
+
+    /**
+     * The resolve function can resolve any suspended function that yields an Either into one type of value.
+     *
+     * @param f the function that needs to be resolved.
+     * @param ifSuccess the function to apply if [f] yields a success of type [A].
+     * @param ifDomainError the function to apply if [f] yields a domain error of type [E].
+     * @param ifSystemFailure the function to apply if [f] throws a [Throwable].
+     * @param ifHandlingCaseThrows the function to apply if the [ifSuccess], [ifDomainError] or [ifSystemFailure] function throw a [Throwable].
+     * Throwing any [Throwable] in the [ifHandlingCaseThrows] function will render the [resolve] function nondeterministic.
+     * @param ifUnrecoverableState the function to apply if [resolve] is in an unrecoverable state.
+     * @return the result of applying the [resolve] function.
+     */
+    suspend fun <E, A, B> resolve(
```

Comment:
can we make all of these inline and add some docs to the Either and Fx docs regarding this new combinator?

Reply:
What exactly are you proposing to inline? 😯 
I'll look into adding some docs soon.

Reply:
Added some documentation. I can add some more documentation in a separate PR to the Arrow Fx docs. Perhaps (also) to the part about arrow-fx-coroutines-kotlinx-coroutines with a specific example of how to use the `resolve` function in Ktor routes.

Reply:
> What exactly are you proposing to inline? 😯 
Hi @joramnv, this is looking great, I ment inlining the resolve function itself.

Reply:
Hi Raúl, thank you. Your proposal to inline the resolve function is for performance reasons, right? Inlining it would mean fewer objects being instantiated and thus a smaller memory footprint and fewer objects that the GC has to deal with. Or are there other reasons to do this as well? (Just so that I understand why to inline. 😄)

I did change the function to be inlined, see last commit. But I had to mark some function parameters as noinline, because else the function could not use default values. Is this a desirable approach? Or should I investigate how to model the same behaviour without using default values (probably using function overloading).

By the way, I also did change the `catch` function to be inlined. I figured that the same applies to this function.

Reply:
I thought some more about it and I think this is a good compromise:
```
    suspend inline fun <E, A, B> resolve(
      f: suspend () -> Either<E, A>,
      success: suspend (a: A) -> Either<Throwable, B>,
      error: suspend (e: E) -> Either<Throwable, B>,
      throwable: suspend (throwable: Throwable) -> Either<Throwable, B>,
      unrecoverableState: suspend (throwable: Throwable) -> Either<Throwable, Unit>
    ): B =
      catch { f() }
        .fold({ t: Throwable -> throwable(t) }, { it.fold({ e: E -> handleItSafely { error(e) } }, { a: A -> handleItSafely { success(a) } }) })
        .fold({ t: Throwable -> throwable(t) }, { b: B -> b.right() })
        .fold({ t: Throwable -> unrecoverableState(t); throw t }, { b: B -> b })
```
The downsides I see are: less versatility (can not supply a function to specifically recover from success, error and throwable failures (but most often it would be the same function as the throwable function anyway)) and there now is a need to always supply an `unrecoverableState` function (which you probably want to do in most production code anyway and if you really want to do nothing at that point, you only need to add `Unit.right()` to the call sight). 

The upside is that the resolve function becomes fully inlined (no more noinline) and will profit from all performance benefits that inline functions bring. Also, if later on we decide that it is desired to have the possibility of a `handlingCaseThrows` like recovering functionality, there would still be a possibility to add an overloaded resolve function with such a function parameter.